### PR TITLE
Filter System

### LINKED ISSUE #6: Profanity Filter
# Profanity Filter

**Features**

* Filter profanity
* Ability to adjust strictness
* See tutorials for possible implementation

### PR DESCRIPTION
**Pull request by MrDogeBro**
* Closes #6

### CODE REVIEW

File: carberretta/utils/chron.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import datetime as dt
 from time import strftime
+from json import JSONEncoder
```
Comment (parafoxia) [inline carberretta/utils/chron.py:3]: Alphabetise imports as per PEP 8 standards.
```suggestion
from json import JSONEncoder
from time import strftime
```

File: carberretta/utils/chron.py [hunk 2]
```diff
@@ -82,3 +83,9 @@ def from_iso(stamp):
 
 def to_iso(obj):
     return obj.isoformat(" ")
+
+
+class DateTimeEncoder(JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, (dt.date, dt.datetime)):
+            return obj.isoformat()
```
Comment (parafoxia) [inline carberretta/utils/chron.py:91]: I would say place this above `sys_time` on line 7.

File: carberretta/bot/cogs/mod.py [hunk 3]
```diff
@@ -13,20 +13,32 @@
...
 import string
 import typing as t
+import json
+import aiofiles
 from collections import defaultdict
+from content_filter import checkMessageList, useCustomListFile, updateListFromFile
+from os import path
 
 import discord
 from discord.ext import commands
```
Comment (parafoxia) [inline carberretta/bot/cogs/mod.py:23]: PEP 8 imports.
```suggestion
import datetime as dt
import json
import string
import typing as t
from collections import defaultdict
from os import path

import aiofiles
import discord
from content_filter import checkMessageList, useCustomListFile, updateListFromFile
from discord.ext import commands
```

File: carberretta/bot/cogs/mod.py [hunk 4]
```diff
@@ -13,20 +13,32 @@
...
 
 from carberretta import Config
-from carberretta.utils import chron
+from carberretta.utils import chron, DEFAULT_EMBED_COLOUR
```
Comment (parafoxia) [inline carberretta/bot/cogs/mod.py:26]:
```suggestion
from carberretta.utils import DEFAULT_EMBED_COLOUR, chron
```

File: carberretta/bot/cogs/mod.py [hunk 5]
```diff
@@ -80,17 +92,153 @@ async def nickname_valid(self, nickname: str) -> bool:
...
+            } # type: dict
+
+            if filter_result_raw:
+                await message.delete()
+
+                for word in filter_result_raw:
+                    filter_result['found'].append(word['word'] + '\n')
+                    filter_result['count'].append(str(word['count']) + '\n')
+
+                warning_msg = await message.channel.send(f"{message.author.mention} please do not use offensive language!")
```
Comment (parafoxia) [inline carberretta/bot/cogs/mod.py:112]:
```suggestion
                warning_msg = await message.channel.send(f"{message.author.mention}, please do not use offensive language.")
```

File: carberretta/bot/cogs/mod.py [hunk 6]
```diff
@@ -80,17 +92,153 @@ async def nickname_valid(self, nickname: str) -> bool:
...
+                member = self.bot.guild.get_member(message.author.id)
+
+                await self.modlog_channel.send(
+                    embed=discord.Embed.from_dict(
+                        {
+                            "title": "Filtered Message",
+                            "color": 0xe33838,
+                            "thumbnail": {"url": f"{member.avatar_url}"},
+                            "footer": {"text": f"ID: {message.id}"},
+                            "image": {"url": att[0].url if len((att := message.attachments)) else None},
```
Comment (parafoxia) [inline carberretta/bot/cogs/mod.py:123]: You don't need the `len(...)` call here.
```suggestion
                            "image": {"url": att[0].url if (att := message.attachments) else None},
```

File: carberretta/bot/cogs/mod.py [hunk 7]
```diff
@@ -80,17 +92,153 @@ async def nickname_valid(self, nickname: str) -> bool:
...
+                            "fields": [
+                                {"name": "Member", "value": member.mention, "inline": False},
+                                {"name": "Message", "value": message.content, "inline": False},
+                                {"name": "Found", "value": '||' + "".join(filter_result['found']) + '||', "inline": True},
+                                {"name": "Count", "value": "".join(filter_result['count']), "inline": True},
+                                {"name": "Context",
+                                    "value": f'[Jump to Message](https://discordapp.com/channels/{message.guild.id}/{message.channel.id}/{warning_msg.id})', "inline": False},
```
Comment (parafoxia) [inline carberretta/bot/cogs/mod.py:130]: `Message` objects have a `jump_url` attribute: https://discordpy.readthedocs.io/en/latest/api.html?highlight=file#discord.Message.jump_url

File: carberretta/bot/cogs/mod.py [hunk 8]
```diff
@@ -80,17 +92,153 @@ async def nickname_valid(self, nickname: str) -> bool:
...
+                        await fix_item_null(index, item, 'added_by', self.bot.user.id)
+                        await fix_item_null(index, item, 'added_on', dt.datetime.utcnow())
+                        await fix_item_null(index, item, 'edited_by', None)
+                        await fix_item_null(index, item, 'edited_on', None)
+
+            for index in indexes_to_remove:
+                filter_data['mainFilter'].pop(index)
+
+            async with aiofiles.open(self.filter_file, "w", encoding="utf-8") as f:
+                await f.write(json.dumps(filter_data, indent=4, cls=chron.DateTimeEncoder))
```
Comment (parafoxia) [inline carberretta/bot/cogs/mod.py:212]: The `indent` kwarg isn't necessary unless the user is going to see the file.
```suggestion
                await f.write(json.dumps(filter_data, cls=chron.DateTimeEncoder))
```

File: carberretta/bot/cogs/mod.py [hunk 9]
```diff
@@ -80,17 +92,153 @@ async def nickname_valid(self, nickname: str) -> bool:
...
+
+        else:
+            filter_file_template = {
+                "mainFilter": [],
+                "dontFilter": None,
+                "conditionFilter": []
+            }
+
+            async with aiofiles.open(self.filter_file, "w", encoding="utf-8") as f:
+                await f.write(json.dumps(filter_file_template, indent=4, cls=chron.DateTimeEncoder))
```
Comment (parafoxia) [inline carberretta/bot/cogs/mod.py:222]: Same as line 212.

File: carberretta/bot/cogs/mod.py [hunk 10]
```diff
@@ -125,6 +273,436 @@ async def validatenicknames_command(self, ctx):
...
+            'added_by': ctx.author.id,
+            'added_on': dt.datetime.utcnow(),
+            'edited_by': None,
+            'edited_on': None
+        }
+
+        filter_data['mainFilter'].append(word_to_add)
+
+        async with aiofiles.open(self.filter_file, "w", encoding="utf-8") as f:
+            await f.write(json.dumps(filter_data, indent=4, cls=chron.DateTimeEncoder))
```
Comment (parafoxia) [inline carberretta/bot/cogs/mod.py:310]: Same as line 212.

File: carberretta/bot/cogs/mod.py [hunk 11]
```diff
@@ -125,6 +273,436 @@ async def validatenicknames_command(self, ctx):
...
+            'added_by': word_found_info['added_by'],
+            'added_on': word_found_info['added_on'],
+            'edited_by': word_found_info['edited_by'],
+            'edited_on': word_found_info['edited_on']
+        }
+
+        filter_data['mainFilter'].remove(word_to_remove)
+
+        async with aiofiles.open(self.filter_file, "w", encoding="utf-8") as f:
+            await f.write(json.dumps(filter_data, indent=4, cls=chron.DateTimeEncoder))
```
Comment (parafoxia) [inline carberretta/bot/cogs/mod.py:347]: Same as line 212.

File: carberretta/bot/cogs/mod.py [hunk 12]
```diff
@@ -125,6 +273,436 @@ async def validatenicknames_command(self, ctx):
...
+            'added_by': word_found_info['added_by'],
+            'added_on': word_found_info['added_on'],
+            'edited_by': ctx.author.id,
+            'edited_on': dt.datetime.utcnow()
+        }
+
+        filter_data['mainFilter'][word_found_index] = word_to_edit
+
+        async with aiofiles.open(self.filter_file, "w", encoding="utf-8") as f:
+            await f.write(json.dumps(filter_data, indent=4, cls=chron.DateTimeEncoder))
```
Comment (parafoxia) [inline carberretta/bot/cogs/mod.py:393]: Same as line 212.

File: carberretta/bot/cogs/mod.py [hunk 13]
```diff
@@ -125,6 +273,436 @@ async def validatenicknames_command(self, ctx):
...
+            'added_by': ctx.author.id,
+            'added_on': dt.datetime.utcnow(),
+            'edited_by': None,
+            'edited_on': None
+        }
+
+        filter_data['conditionFilter'].append(word_to_add)
+
+        async with aiofiles.open(self.filter_file, "w", encoding="utf-8") as f:
+            await f.write(json.dumps(filter_data, indent=4, cls=chron.DateTimeEncoder))
```
Comment (parafoxia) [inline carberretta/bot/cogs/mod.py:431]: Same as line 212.

File: carberretta/bot/cogs/mod.py [hunk 14]
```diff
@@ -125,6 +273,436 @@ async def validatenicknames_command(self, ctx):
...
+            'added_by': word_found_info['added_by'],
+            'added_on': word_found_info['added_on'],
+            'edited_by': word_found_info['edited_by'],
+            'edited_on': word_found_info['edited_on']
+        }
+
+        filter_data['conditionFilter'].remove(word_to_remove)
+
+        async with aiofiles.open(self.filter_file, "w", encoding="utf-8") as f:
+            await f.write(json.dumps(filter_data, indent=4, cls=chron.DateTimeEncoder))
```
Comment (parafoxia) [inline carberretta/bot/cogs/mod.py:468]: Same as line 212.

File: carberretta/bot/cogs/mod.py [hunk 15]
```diff
@@ -125,6 +273,436 @@ async def validatenicknames_command(self, ctx):
...
+            'added_by': word_found_info['added_by'],
+            'added_on': word_found_info['added_on'],
+            'edited_by': ctx.author.id,
+            'edited_on': dt.datetime.utcnow()
+        }
+
+        filter_data['conditionFilter'][word_found_index] = word_to_edit
+
+        async with aiofiles.open(self.filter_file, "w", encoding="utf-8") as f:
+            await f.write(json.dumps(filter_data, indent=4, cls=chron.DateTimeEncoder))
```
Comment (parafoxia) [inline carberretta/bot/cogs/mod.py:514]: Same as line 212.

**Review by parafoxia — CHANGES_REQUESTED** [review]
Good PR! Overall looks pretty sound, most suggestions made are PEP 8 standards, etc.

However, you do need to update the dependencies and the config files. Run `poetry add content_filter` in the terminal to add that dependency, and add any extra config IDs (such as the `MODLOG_ID`) to the docker-compose.yml file.

File: carberretta/bot/cogs/mod.py (hunk 7 above)
Comment (MrDogeBro) [reply carberretta/bot/cogs/mod.py:130]: Thanks for that info — looked around but couldn't find this but there it is

File: carberretta/bot/cogs/mod.py (hunk 6 above)
Comment (MrDogeBro) [reply carberretta/bot/cogs/mod.py:123]: Fixed where it was copied from your code above as well

**Comment by MrDogeBro** [comment]
Ok all changes should be updated as per request. Wouldn't be a bad idea for you to go through and check that they are all in place how you would like them to be though.

**Comment by MrDogeBro** [comment]
Just added the ability for the user to specify if a word in the conditional filter needs to have a space in front of the word or not for the conditional filter to pick up on the word.

**Review by parafoxia — APPROVED** [review]
Just the one really simple thing, but I'm pretty sure I can commit it on my end. Thanks for the contribution!

**Review by parafoxia — DISMISSED** [review]

**Review by mshaugh — CHANGES_REQUESTED** [review]
Please revert changes made to:

* `carberretta/bot/cogs/support.py`
* `carberretta/bot/cogs/youtube.py`
* `carberretta/bot/utils/string.py`

In the same vein, please also clean up the git history by removing some of the 'useless' commits (e.g. 
23b1426).

Additionally, please undo the version change in `pyproject.toml`.

To do this:

```shell
$ git checkout filterSystem
$ git rebase -i HEAD~n # where n is the number of last commits you want to include in interactive rebase.
# Replace pick with drop for commits you want to discard.
# Save and exit.
$ git push --force
```

To be clear, the pull request should only contain changes directly related to the feature: its documentation, implementation, dependencies, and so forth.

As for `carberretta/bot/cogs/mod.py`:

* Prefer the use of [`pathlib`](https://docs.python.org/3/library/pathlib.html) over `os.path`
* Use [`str.translate`](https://docs.python.org/3/library/stdtypes.html#str.translate) instead of long `replace` chains
* I don't see the necessity of the `fix_x_y` functions in `load_filter_file`
* Much of the logic can be inverted to improve readability
    ```python
    # Bad
    if condition:
        do_something()
    else:
        error()

    # Good:
    if not condition:
        error()
    
    do_something()
    ```
* Use [`del`](https://docs.python.org/3/reference/simple_stmts.html#the-del-statement), rather than creating a list of keys to remove
* Generalise 'word_found' logic to a single function
    * Currently, it is repeated across many functions with slight differences. It shouldn't be too hard to generalise, returning the found word or raising appropriate exceptions.
* It is trivial to bypass the filter
    * Non-English characters (e.g. [ϲ](https://en.wikipedia.org/wiki/Sigma#Lunate_sigma), [ç](https://en.wikipedia.org/wiki/%C3%87) instead of [c](https://en.wikipedia.org/wiki/C))
    * Non-printing characters (e.g. [ZWSP](https://en.wikipedia.org/wiki/Zero-width_space))


As an aside, and this is most certainly not a required change, please consider patching [content_filter](https://pypi.org/project/content-filter/) to be more PEP8 compliant. Specifically [function names](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names) should be snake case (lowercase_with_underscores)

**Comment by MrDogeBro** [comment]
- [ ] Update to use `pathlib` over `os.path`
- [ ] Use `str.translate` over replace chains
- [ ] Generalise 'word_found' logic to a single function
- [ ] Remove useless commits
- [ ] Invert logic for readability
- [ ] Improve filtering in `content_filter` module
  - [Non-English](https://github.com/MrDogeBro/content_filter/issues/2) charaters
  - [Non-printing](https://github.com/MrDogeBro/content_filter/issues/3) charaters
- [ ] Make `content_filter` [comply with PEP8](https://github.com/MrDogeBro/content_filter/issues/4)

**Comment by MrDogeBro** [comment]
Ok, so changes have been made now. I think that everything you wanted @mshaugh has been changed. If either of you two see anything I missed just tell me and I'll fix it. If you'd like anything else changed just tell me as well. Also sorry for the long delay. I've been busy and working on other stuff recently.

**Comment by parafoxia** [comment]
It's okay, we've been busy too. We still need to review Carberretta's infrastructure for the v2 release, so we'll review the changes after that's complete.

**Comment by MrDogeBro** [comment]
The feature has been added so that an automod action can be specified. The default action is just a warning but you can set an individual word to be a warn, kick, or ban when detected. Currently it is just a verbal warning but in the future this will integrate with the new moderation system for Carberretta v2 so that the warnings actually count and the kick and ban actually kick or ban the person. When the new moderation system is introduced, I can help with integrating the filter auto mod into it if needed/wanted.

**Comment by MrDogeBro** [comment]
I added a verbal warn action and the verbal warn action is now the default if an action is not provided. This will integrate into the new verbal/zero warn system for Carberretta in the future.

**Comment by parafoxia** [comment]
Superceded by #137.

**Pull request closed without merging**